File: examples/offline_signing/0_address_generation.rs
// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! In this example we generate addresses which will be used later to find inputs.
//! This example uses dotenv, which is not safe for use in production.
//! `cargo run --example 0_address_generation --release`.

use std::{
    fs::File,
    io::{BufWriter, Write},
    path::Path,
};

use iota_client::{
    constants::SHIMMER_TESTNET_BECH32_HRP,
    secret::{mnemonic::MnemonicSecretManager, SecretManager},
    Client, Result,
};

const ADDRESS_FILE_NAME: &str = "examples/offline_signing/addresses.json";

#[tokio::main]
async fn main() -> Result<()> {
    dotenv::dotenv().ok();

    // Creates a client instance.
    let offline_client = Client::builder().with_offline_mode().finish()?;
    let secret_manager = SecretManager::Mnemonic(MnemonicSecretManager::try_from_mnemonic(
        &std::env::var("NON_SECURE_USE_OF_DEVELOPMENT_MNEMONIC_1").unwrap(),
    )?);

    // Generates addresses offline.
    let addresses = offline_client
        .get_addresses(&secret_manager)
        .with_range(0..10)
        .with_bech32_hrp(SHIMMER_TESTNET_BECH32_HRP)
        .finish()
        .await?;

    write_addresses_to_file(ADDRESS_FILE_NAME, &addresses)
}

fn write_addresses_to_file<P: AsRef<Path>>(path: P, addresses: &[String]) -> Result<()> {
    let json = serde_json::to_string_pretty(&addresses)?;
    let mut file = BufWriter::new(File::create(path)?);

    println!("{}", json);

    file.write_all(json.as_bytes())?;

    Ok(())
}

File: bindings/nodejs/examples/offline_signing/0_address_generation.js
// Copyright 2021-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// In this example we will generate addresses which will be used later to find inputs
async function run() {
    const {
        Client,
        initLogger,
        SHIMMER_TESTNET_BECH32_HRP,
    } = require('@iota/client');
    const { writeFile } = require('fs/promises');
    require('dotenv').config({ path: '../.env' });

    initLogger({
        colorEnabled: true,
        name: './client.log',
        levelFilter: 'debug',
    });

    // client will connect to testnet by default
    const offlineClient = new Client({
        offline: true,
        nodes: [
            {
                // Insert your node URL here.
                url: 'http://localhost:14265/',
                auth: null,
                disabled: false,
            },
        ],
        localPow: true,
    });

    const signer = JSON.stringify({
        Mnemonic: process.env.NON_SECURE_USE_OF_DEVELOPMENT_MNEMONIC_1,
    });
    const ADDRESS_FILE_NAME = './addresses.json';

    const offlineGeneratedOptions = {
        range: {
            start: 0,
            end: 10,
        },
        bech32Hrp: SHIMMER_TESTNET_BECH32_HRP,
    };

    try {
        // Generates addresses offline.
        const offlineGeneratedAddresses = await offlineClient.generateAddresses(
            signer,
            offlineGeneratedOptions,
        );

        await writeFile(
            ADDRESS_FILE_NAME,
            JSON.stringify(offlineGeneratedAddresses),
            (err) => {
                if (err) {
                    console.error(err);
                }
            },
        );
    } catch (error) {
        console.error('Error: ', error);
    }
}

run().then(() => process.exit());
